src/assets/scripts/client/trafficGenerator/SpawnScheduler.js

```diff
@@ -89,20 +108,63 @@ export default class SpawnScheduler {
      * @for SpawnScheduler
      * @method createNextSchedule
      * @param spawnPatternModel {SpawnPatternModel}
-     * @param aircraftController {AircraftTypeDefinitionCollection}
-     * @return {function}
+     * @return {array}
      */
-    createNextSchedule(spawnPatternModel, aircraftController) {
+    createNextSchedule(spawnPatternModel) {
         const delay = spawnPatternModel.getNextDelayValue(TimeKeeper.accumulatedDeltaTime);
 
+        return this._createTimeout(spawnPatternModel, delay);
+    }
+
+    /**
+     * Resets the timer for a specific spawn pattern
+     *
+     * @for SpawnScheduler
+     * @method resetTimer
+     * @param spawnPatternModel {SpawnPatternModel}
+     */
+    resetTimer(spawnPatternModel) {
+        let timePassed = 0;
+
+        if (spawnPatternModel.scheduleId) {
+            GameController.destroyTimer(spawnPatternModel.scheduleId);
+
+            const timerStart = spawnPatternModel.scheduleId[1] - spawnPatternModel.scheduleId[3];
+            timePassed = TimeKeeper.accumulatedDeltaTime - timerStart;
+            spawnPatternModel.scheduleId = null;
+        }
+
+        if (spawnPatternModel.rate > 0) {
+            let nextDelay = spawnPatternModel.getNextDelayValue(TimeKeeper.accumulatedDeltaTime);
+
+            if (timePassed < nextDelay) {
+                nextDelay -= timePassed;
+            } else {
+                this.aircraftController.createAircraftWithSpawnPatternModel(spawnPatternModel);
+            }
+
+            spawnPatternModel.scheduleId = this._createTimeout(spawnPatternModel, nextDelay);
+        }
+    }
+
+    /**
+     * Registers a new timeout, its callback and callback arguments with the `GameController`
+     *
+     * @for SpawnScheduler
+     * @method _createTimeout
+     * @param spawnPatternModel {SpawnPatternModel}
+     * @param delay {number} time in seconds
+     * @return {array}
+     */
+    _createTimeout(spawnPatternModel, delay) {
         return GameController.game_timeout(
             this.createAircraftAndRegisterNextTimeout,
             // lifespan of timeout
             delay,
             // passing null only to match existing api
             null,
             // arguments sent to callback as it's first parameter. using array so multiple arg can be sent
-            [spawnPatternModel, aircraftController]
+            [spawnPatternModel, this.aircraftController]
```

Comment:
does this reference stay the same when called from the `.game_timeout()`?  Meaning, the context of `this` is correct when it's called from `.game_timeout()`?

Reply:
yes